### PR TITLE
added nodeAffinity during notebook creation with 0 gpus

### PR DESCRIPTION
**Pull request by mlassak**
Resolves: #381 

## Description
Added soft nodeAffinity when the number of GPUs is zero.

image

## How Has This Been Tested?
1. Deploy this PR on the cluster
2. Spawn notebook with 0 GPUs
3. Verify that posted notebook has the desired affinity
4. Attempt spawn with at least 1 GPU
5. Verify that the affinity is not present

## Merge criteria:
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: frontend/src/services/notebookService.ts [hunk 1]
```diff
@@ -37,6 +37,7 @@ export const createNotebook = (
   const url = `/api/notebooks/${projectName}`;
   const resources = { ...notebookSize?.resources };
   const tolerations: NotebookToleration[] = [];
+  let affinity = {};
```
Comment (maroroman) [inline frontend/src/services/notebookService.ts:40]: @samuelvl Is an empty affinity object okay for the notebook pod if gpus are present?

**Review by maroroman — APPROVED** [review]
lgtm

File: frontend/src/services/notebookService.ts (hunk 1 above)
Comment (samuelvl) [reply frontend/src/services/notebookService.ts:40]: Yes, it's totally fine

**Comment by andrewballantyne** [comment]
@samuelvl can you review this again?

**Review by samuelvl — APPROVED** [review]
I confirm the affinity is added when no GPU is selected:

```yaml
$ oc get sts jupyter-nb-user-40redhat-2ecom -o yaml
spec:
  template:
    spec:
      affinity:
        nodeAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - preference:
              matchExpressions:
              - key: nvidia.com/gpu.present
                operator: NotIn
                values:
                - "true"
            weight: 1
```

And no affinity is added when GPU is selected.

/lgtm

**Comment by samuelvl** [comment]
/hold

@Xaenalt Do we want this feature in RHODS 1.16 or 1.17? I see in RHODS-3074 the target release is 1.17 but the PR is ready to be merged for 1.16.

**Comment by Xaenalt** [comment]
I'm fine with it in 1.16

**Comment by lugi0** [comment]
@samuelvl @Xaenalt I see the affinity also on pods that have requested a GPU, is this expected?

```
[lgiorgi@fedora ods-ci]$ oc exec jupyter-nb-ldap-2dadmin4-0 -n rhods-notebooks  -- nvidia-smi
Defaulted container "jupyter-nb-ldap-2dadmin4" out of: jupyter-nb-ldap-2dadmin4, oauth-proxy
Tue Aug 30 09:08:06 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
| N/A   34C    P8    15W /  70W |      0MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
[lgiorgi@fedora ods-ci]$ oc get sts jupyter-nb-ldap-2dadmin4 -o yaml -n rhods-notebooks | yq -y .spec.template.spec.affinity
nodeAffinity:
  preferredDuringSchedulingIgnoredDuringExecution:
    - preference:
        matchExpressions:
          - key: nvidia.com/gpu.present
            operator: NotIn
            values:
              - 'true'
      weight: 1
```

**Comment by lugi0** [comment]
(for completeness, when there's no GPU)
```
[lgiorgi@fedora ods-ci]$ oc exec jupyter-nb-ldap-2dadmin4-0 -n rhods-notebooks  -- nvidia-smi
Defaulted container "jupyter-nb-ldap-2dadmin4" out of: jupyter-nb-ldap-2dadmin4, oauth-proxy
time="2022-08-30T09:11:06Z" level=error msg="exec failed: unable to start container process: exec: \"nvidia-smi\": executable file not found in $PATH"
command terminated with exit code 255
[lgiorgi@fedora ods-ci]$ oc get sts jupyter-nb-ldap-2dadmin4 -o yaml -n rhods-notebooks | yq -y .spec.template.spec.affinity
nodeAffinity:
  preferredDuringSchedulingIgnoredDuringExecution:
    - preference:
        matchExpressions:
          - key: nvidia.com/gpu.present
            operator: NotIn
            values:
              - 'true'
      weight: 1
```

**Comment by maroroman** [comment]
Ran a quick test with a local image build and it seems there is no affinity in the POST call that creates the notebook, however checking on the actual created object after the POST call is complete results in the above. So the dashboard frontend or backend does not ask for the affinity, but it appears anyway.

**Comment by Xaenalt** [comment]
So, the intent would be to only apply the affinity when there's no GPU explicitly requested, however it doesn't break anything to have it in either case. The GPU overrides the soft affinity, so it'll still spawn correctly

**Pull request merged**